File: EVR/src/main/java/com/group7/evr/service/BookingService.java
package com.group7.evr.service;

import com.group7.evr.entity.Booking;
import com.group7.evr.entity.Contract;
import com.group7.evr.entity.User;

import java.util.List;
import java.util.Map;

public interface BookingService {
    Booking createBooking(Booking booking, User user);

    Booking getBookingById(Integer bookingId);

    Map<String, Object> getUserBookingsWithFilters(Integer userId, String status, String fromDate, String toDate, int page, int size);
        
    Booking checkIn(Integer bookingId, User user, User staff);

    Booking returnVehicle(Integer bookingId, User user, User staff, Double batteryLevel);

    List<Booking> getUserHistory(Integer userId);

    List<Booking> getCheckInQueue(Integer staffId);

    List<Booking> getReturnQueue(Integer staffId);

    List<Booking> getStaffContracts(Integer staffId);

    Map<String, Object> getUserAnalytics(Integer userId);

    Map<String, Object> getAdvancedUserAnalytics(Integer userId);

    Booking modifyBooking(Integer bookingId, Booking updates, User actor);

    Booking cancelBooking(Integer bookingId, User actor);

    Booking denyBooking(Integer bookingId, User staff, String reason);

    Map<String, Object> settleBooking(Integer bookingId, User actor);
    Map<String, Object> getStaffContractsWithDetails(Integer staffId);
}

File: EVR-FE/src/services/booking.service.js
import apiClient from './api/axios';
import { ENDPOINTS } from './api/endpoints';

export const bookingService = {
  createBooking: async (bookingData, userId) => {
    return apiClient.post(ENDPOINTS.CREATE_BOOKING, bookingData, {
      params: { userId },
    });
  },

  getBooking: async (id) => {
    return apiClient.get(ENDPOINTS.GET_BOOKING(id));
  },

  checkIn: async (id, userId, staff) => {
    return apiClient.put(ENDPOINTS.CHECKIN_BOOKING(id), staff, {
      params: { userId },
    });
  },

  returnVehicle: async (id, userId, staff) => {
    return apiClient.put(ENDPOINTS.RETURN_BOOKING(id), staff, {
      params: { userId },
    });
  },

  getUserBookings: async (userId) => {
    return apiClient.get(ENDPOINTS.GET_USER_BOOKINGS, {
      params: { userId },
    });
  },

  getUserBookingsAdvanced: async (userId, status = null, fromDate = null, toDate = null, page = 0, size = 10) => {
    const params = { page, size };
    if (status) params.status = status;
    if (fromDate) params.fromDate = fromDate;
    if (toDate) params.toDate = toDate;
    return apiClient.get(ENDPOINTS.GET_USER_BOOKINGS_ADVANCED(userId), { params });
  },

  modifyBooking: async (id, updates, userId) => {
    return apiClient.put(ENDPOINTS.MODIFY_BOOKING(id), updates, {
      params: { userId },
    });
  },

  cancelBooking: async (id, userId) => {
    return apiClient.delete(ENDPOINTS.CANCEL_BOOKING(id), {
      params: { userId },
    });
  },

  denyBooking: async (id, staffId, reason = null) => {
    return apiClient.put(ENDPOINTS.DENY_BOOKING(id), reason ? { reason } : null, {
      params: { staffId },
    });
  },

  settleBooking: async (id, userId) => {
    return apiClient.post(ENDPOINTS.SETTLE_BOOKING(id), null, {
      params: { userId },
    });
  },

  getUserAnalytics: async (userId) => {
    return apiClient.get(ENDPOINTS.GET_USER_ANALYTICS, {
      params: { userId },
    });
  },

  getAdvancedAnalytics: async (userId) => {
    return apiClient.get(ENDPOINTS.GET_ADVANCED_ANALYTICS, {
      params: { userId },
    });
  },

  returnInspection: async (id, userId) => {
    return apiClient.post(ENDPOINTS.RETURN_INSPECTION(id), null, {
      params: { userId },
    });
  },

  getStaffCheckInQueue: async (staffId) => {
    return apiClient.get(ENDPOINTS.STAFF_CHECKIN_QUEUE, {
      params: { staffId },
    });
  },

  getStaffReturnQueue: async (staffId) => {
    return apiClient.get(ENDPOINTS.STAFF_RETURN_QUEUE, {
      params: { staffId },
    });
  },

  getStaffContracts: async (staffId) => {
    return apiClient.get(ENDPOINTS.STAFF_CONTRACTS, {
      params: { staffId },
    });
  },

  getContractByBookingId: async (bookingId) => {
    return apiClient.get(ENDPOINTS.STAFF_CONTRACT_BY_BOOKING(bookingId));
  },
};

